docs/src/interface-definition.md

```diff
@@ -0,0 +1,227 @@
+- `AtomsCalculators.calculate` using `AtomsCalculators.Virial()` as the first argument for virial calculation
+
+You do not need to implement all of these by yourself. There is macro that will help implement the other calls. 
+
+Each of the individual calls have two common inputs: `AtomsBase.AbstractSystem` compatible structure and a `calculator` that incudes details of the calculation method. Calculate interface has additionally the type of calculation as the first input. You can tune calculation by passing keyword arguments, which can be ignored, but they need to be present in the function definition.
+
+`potential_energy`, `forces`, `forces!` and `virial`:
+
+- First input is `AtomsBase.AbstractSystem` compatible structure
+- Second input is `calculator` structure
+- Method has to accept keyword arguments (they can be ignored)
+- Non-allocating force call `force!` has an AbstractVector as the first input, to which the evaluated force values are stored (look for more details below)
+
+`calculate`:
+
+- First input is either `Energy()`, `Forces()` or `Virial()`
+
+Outputs for the functions need to have following properties
+
+- Energy is a subtype of `Number` that has a unit with dimensions of energy (mass * length^2 / time^2)
+- Force output is a subtype of `AbstractVector` with element type also a subtype of AbstractVector (length 3 in 3D) and unit with dimensions of force (mass * length / time^2). With additional property that it can be reinterpret as a matrix
+- Virial is a square matrix (3x3 in 3D) that has units of force times length or energy
+
+
+## Implementing the interface
+
+You can either implement both of the calls e.g. for energy
+
+`AtomsCalculators.potential_energy(system, calculator; kwargs...)` and
+`AtomsCalculators(AtomsCalculators.Energy(), system, calculator; kwargs...)`
+
+### Example implementations
+
+Example `potential_energy` implementation
+
+```julia
+using AtomsCalculators
+using Unitful
+struct MyType
+end
+
+AtomsCalculators.@generate_interface function AtomsCalculators.potential_energy(system, calculator::Main.MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition here
+    return 0.0u"eV"
+end
+```
+
+!!! note "Type definition under @generate_complement macro"
+    You need to use explicit definition of type when using
+    `@generate_interface` macro. `Main.MyType` is fine `MyType` is not!
+
+    You also need to define the type before the macro call.
+
+Completely equivalent implementation is
+
+```julia
+AtomsCalculators.@generate_interface function AtomsCalculators.calculate(::AtomsCalculators.Energy, system, calculator::Main.MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition here
+    return 0.0u"eV"
```

Comment:
I think here I would also always enforce returning a named tuple, since this is the low-level interface. This gives us more flexibility later without breaking anything (e.g. if the user requests multiple things to be calculated, if state is passed around etc.).

Reply:
NamedTuple output is now in place and forced by the tests. The test will check for keys on the output and expects `:energy`, `:forces` or `:virial` to be present.